File: Striver_SDE_SHEET/891-score-after-flipping-matrix/score-after-flipping-matrix.cpp
class Solution {
public:
    int matrixScore(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        // row optimization
        for(int i=0;i<m;i++){
            if(grid[i][0]==0){
                for(int j=0;j<n;j++){
                    grid[i][j] = 1-grid[i][j];
                }
            }
        }

        // optimize columns
        for(int j=0;j<n;j++){
            int count = 0;
            for(int i=0;i<m;i++){
                count+=grid[i][j]==0?1:0;
            }
            if(count>m-count){
                for(int i=0;i<m;i++){
                    grid[i][j] = 1-grid[i][j];
                }
            }
        }

        int score = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int columnScore = grid[i][j] << (n - j - 1);
                score += columnScore;
            }
        }
        return score;
    }
};

File: Striver_SDE_SHEET/891-score-after-flipping-matrix/score-after-flipping-matrix.py
class Solution:
    def matrixScore(self, grid: List[List[int]]) -> int:
        m,n = len(grid),len(grid[0])
        rev_row = lambda i: [1-grid[i][j] for j in range(n)]
        
        for i in range(m):
            if grid[i][0]==0: grid[i] = rev_row(i)
        # print(grid)
        for i in range(1,n):
            countz = 0
            for j in range(m):
                countz += int(grid[j][i]==0)
            if m-countz < countz:
                for j in range(m):
                    grid[j][i] = 1-grid[j][i]
        print(grid)
        ans = 0
        for i in range(m):
            num = 0
            for j in range(n):
                num = (num<<1)
                num = num|grid[i][j]
            ans+=num
        return ans